src/Sparrow/Utils/ZstdStream.cs

```diff
@@ -0,0 +1,361 @@
+        public override long Position { get => throw new NotSupportedException(); set => throw new NotSupportedException(); }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        private unsafe int DecompressStep(ReadOnlySpan<byte> buffer)
+        {
+            fixed (byte* pBuffer = buffer, pOutput = _decompressionInput.Span)
+            {
+                var output = new ZstdLib.ZSTD_outBuffer { Source = pBuffer, Position = UIntPtr.Zero, Size = (UIntPtr)buffer.Length };
+                var input = new ZstdLib.ZSTD_inBuffer { Source = pOutput, Position = UIntPtr.Zero, Size = (UIntPtr)_decompressionInput.Length };
+                var v = ZstdLib.ZSTD_decompressStream(_compressContext.Streaming, &output, &input);
+                ZstdLib.AssertZstdSuccess(v);
+                _decompressionInput = _decompressionInput.Slice((int)input.Position);
+                return (int)output.Position;
+            }
+        }
+
+        private unsafe (int OutputPosition, int InputPosition, bool Done) CompressStep(ReadOnlySpan<byte> buffer, ZstdLib.ZSTD_EndDirective directive)
+        {
+            fixed (byte* pBuffer = buffer, pTempBuffer = _tempBuffer)
+            {
+                var input = new ZstdLib.ZSTD_inBuffer { Source = pBuffer, Position = UIntPtr.Zero, Size = (UIntPtr)buffer.Length };
+                var output = new ZstdLib.ZSTD_outBuffer { Source = pTempBuffer, Position = UIntPtr.Zero, Size = (UIntPtr)_tempBuffer.Length };
+                var v = ZstdLib.ZSTD_compressStream2(_compressContext.Compression, &output, &input, directive);
+                ZstdLib.AssertZstdSuccess(v);
+                return ((int)output.Position, (int)input.Position, v == UIntPtr.Zero);
+            }
+        }
+
+        private unsafe void ShiftBufferData()
+        {
+            if (_decompressionInput.Length == 0)
+                return;
+
+            if (_decompressionInput.Length == _tempBuffer.Length)
+                throw new InvalidOperationException("Should never happen, the buffer is full of data that produces not output");
+
+            fixed (byte* pTempBuf = _tempBuffer, pCurVal = _decompressionInput.Span)
+            {
+                if (pTempBuf == pCurVal)
+                    return;
+                Buffer.MemoryCopy(pCurVal, pTempBuf, _tempBuffer.Length, _decompressionInput.Length);
+                _decompressionInput = new Memory<byte>(_tempBuffer, 0, _decompressionInput.Length);
+            }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            return Read(new Span<byte>(buffer, offset, count));
+        }
+
+        public override int Read(Span<byte> buffer)
+        {
+            if (_readSemaphore.Wait(10_000) == false)
```

Comment:
Time should be configurable, and much smaller by default (1,500 ms)

Reply:
That said, we only need to worry about this for the compression context itself, no?
In that case, let's move the scope of the semaphore to the compression / decompress step.
They are bounded and do not involve I/O, so that is much better.

Reply:
Done